File: backend/app/api/v1/production.py
"""Production workspace index and explicitly requested local diagnostics."""

import json
from typing import Any, Literal

from fastapi import APIRouter, Depends, Query, Request
from pydantic import Field, JsonValue

from app.api.v1.models import Uuid
from app.api.v1.projects import ProjectStore, project_session, window_identity
from app.api.v1.projects_models import BusinessModel, Command, Envelope, MutationReceipt
from app.storage import issues


class DiagnosticOptions(BusinessModel):
    project_id: Uuid | None
    include_project: bool
    include_content: bool


class DiagnosticPayload(DiagnosticOptions):
    target_grant_id: Uuid


class DiagnosticFile(BusinessModel):
    name: str
    description: str


class DiagnosticPreview(DiagnosticOptions):
    files: list[DiagnosticFile]


class DraftSummary(BusinessModel):
    id: Uuid
    artifact_id: Uuid
    kind: Literal["story", "asset", "shot", "speech", "subtitle", "timeline", "observation"]
    saved_at: str


class AdoptedSummary(BusinessModel):
    artifact_id: Uuid
    revision_id: Uuid
    kind: Literal["story", "asset", "shot", "speech", "subtitle", "timeline", "observation"]
    confirmed: bool
    needs_update: bool
    content: dict[str, JsonValue]


class ProductionIndex(BusinessModel):
    drafts: list[DraftSummary] = Field(max_length=5000)
    adopted: list[AdoptedSummary] = Field(max_length=5000)
    next_offset: int | None


class Issue(BusinessModel):
    id: Uuid
    rule_id: str
    rule_version: str
    artifact_id: Uuid
    revision_id: Uuid
    baseline_revision_id: Uuid | None
    severity: Literal["blocking", "unknown_required", "deviation", "advice"]
    status: Literal["open", "fixing", "recheck", "resolved", "accepted_deviation"]
    message: str
    evidence: str
    time: dict[str, int] | None
    method: Literal["local", "ai", "human"]
    limitations: str


class IssuePage(BusinessModel):
    items: list[Issue]
    next_cursor: Uuid | None


class IssueDecision(BusinessModel):
    action: Literal["accept_deviation", "request_recheck", "attach_evidence"]
    reason: str = Field(min_length=1, max_length=2000)
    evidence_media_ids: list[Uuid] = Field(max_length=1000)


def production_router(store: ProjectStore) -> APIRouter:
    router = APIRouter(prefix="/api/v1", tags=["Production workspace and T13 diagnostics"])

    @router.get("/projects/{project_id}/issues", response_model=Envelope[IssuePage])
    def list_issues(
        project_id: Uuid,
        request: Request,
        cursor: Uuid | None = None,
        limit: int = Query(default=50, ge=1, le=200),
        window: int = Depends(window_identity),
    ) -> Any:
        owner = store.get()
        with owner._mutex:
            data = owner._versions.read(
                project_id,
                project_session(request),
                window,
                lambda db, _: issues.page(db, cursor, limit),
            )
        return {"requestId": request.state.request_id, "data": data}

    @router.post(
        "/projects/{project_id}/issues/{issue_id}/decisions",
        response_model=Envelope[MutationReceipt],
    )
    def decision(
        project_id: Uuid,
        issue_id: Uuid,
        command: Command[IssueDecision],
        request: Request,
        window: int = Depends(window_identity),
    ) -> Any:
        owner = store.get()
        with owner._mutex:
            data = owner._versions.write(
                project_id,
                project_session(request),
                window,
                command.model_dump(by_alias=True),
                "decideIssue",
                {"action", "reason", "evidenceMediaIds"},
                issue_id,
                lambda db, _, payload: issues.decide(
                    db, issue_id, payload, command.client_operation_id
                ),
            )
        return {"requestId": request.state.request_id, "data": data}

    @router.get("/projects/{project_id}/production", response_model=Envelope[ProductionIndex])
    def index(
        project_id: Uuid,
        request: Request,
        offset: int = Query(default=0, ge=0),
        window: int = Depends(window_identity),
    ) -> Any:
        owner = store.get()

        def read(db: Any, _: Any) -> Any:
            entries = db.execute(
                "SELECT 'draft' AS category,id FROM drafts UNION ALL "
                "SELECT 'adopted',id FROM artifacts WHERE adopted_revision_id IS NOT NULL "
                "ORDER BY category,id LIMIT 101 OFFSET ?",
                (offset,),
            ).fetchall()
            draft_ids = {r["id"] for r in entries if r["category"] == "draft"}
            artifact_ids = {r["id"] for r in entries if r["category"] == "adopted"}
            drafts = [
                {
                    "id": r["id"],
                    "artifactId": r["artifact_id"],
                    "kind": r["kind"],
                    "savedAt": r["saved_at"],
                }
                for r in db.execute(
                    f"SELECT * FROM drafts WHERE id IN ({','.join('?' for _ in draft_ids)})",
                    tuple(draft_ids),
                )
            ]
            adopted = [
                {
                    "artifactId": r["artifact_id"],
                    "revisionId": r["id"],
                    "kind": r["kind"],
                    "confirmed": r["confirmed_revision_id"] == r["id"],
                    "needsUpdate": bool(r["needs_update"]),
                    "content": json.loads(r["payload_json"])["content"],
                }
                for r in db.execute(
                    "SELECT a.kind,a.confirmed_revision_id,a.needs_update,r.* FROM artifacts a "
                    "JOIN revisions r ON r.id=a.adopted_revision_id "
                    f"WHERE a.id IN ({','.join('?' for _ in artifact_ids)})",
                    tuple(artifact_ids),
                )
            ]
            lookup = {("draft", r["id"]): r for r in drafts}
            lookup.update({("adopted", r["artifactId"]): r for r in adopted})
            page: dict[str, Any] = {"drafts": [], "adopted": [], "nextOffset": None}
            size = 100
            consumed = 0
            for entry in entries[:100]:
                item = lookup[(entry["category"], entry["id"])]
                item_size = len(json.dumps(item, ensure_ascii=False).encode("utf-8"))
                if consumed and size + item_size > 900 * 1024:
                    break
                page["drafts" if entry["category"] == "draft" else "adopted"].append(item)
                size += item_size
                consumed += 1
            if consumed < len(entries):
                page["nextOffset"] = offset + consumed
            return page

        with owner._mutex:
            data = owner._versions.read(project_id, project_session(request), window, read)
        return {"requestId": request.state.request_id, "data": data}

    @router.post("/diagnostics/preview", response_model=Envelope[DiagnosticPreview])
    def preview(
        payload: DiagnosticOptions, request: Request, window: int = Depends(window_identity)
    ) -> Any:
        return {
            "requestId": request.state.request_id,
            "data": store.get()._diagnostics.preview(
                window,
                payload.model_dump(by_alias=True),
                project_session(request) if payload.include_project else None,
            ),
        }

    @router.post("/diagnostics", status_code=202, response_model=Envelope[MutationReceipt])
    def diagnostic(
        command: Command[DiagnosticPayload],
        request: Request,
        window: int = Depends(window_identity),
    ) -> Any:
        return {
            "requestId": request.state.request_id,
            "data": store.get()._diagnostics.create(
                window,
                command.model_dump(by_alias=True),
                project_session(request) if command.payload.include_project else None,
            ),
        }

    return router

File: frontend/electron/shared/production.ts
import { z } from "zod";
import { projectUuid, receiptSchema, type ProjectResult } from "./projects";

const integer = z.number().int().min(0).max(Number.MAX_SAFE_INTEGER);
const text = z.string().max(4000);
const object = z.record(z.string(), z.json());
const kind = z.enum([
  "story",
  "asset",
  "shot",
  "speech",
  "subtitle",
  "timeline",
  "observation",
]);
const project = z.strictObject({ projectId: projectUuid });
const ids = z.array(projectUuid).max(1000);
const hash = z.string().regex(/^[0-9a-f]{64}$/);
export const timelineSchema = z.strictObject({
  width: z.union([
    z.literal(720),
    z.literal(1080),
    z.literal(1280),
    z.literal(1920),
  ]),
  height: z.union([
    z.literal(720),
    z.literal(1080),
    z.literal(1280),
    z.literal(1920),
  ]),
  fps: z.strictObject({
    numerator: z.union([z.literal(24), z.literal(25), z.literal(30)]),
    denominator: z.literal(1),
  }),
  durationMs: integer.min(1),
  burnSubtitles: z.boolean(),
  tracks: z
    .array(
      z.strictObject({
        id: projectUuid,
        kind: z.enum(["image", "video", "voice", "music", "sfx", "subtitle"]),
        order: integer.max(100),
        muted: z.boolean(),
      }),
    )
    .min(1)
    .max(100),
  clips: z
    .array(
      z.strictObject({
        id: projectUuid,
        trackId: projectUuid,
        shotId: projectUuid.nullable(),
        mediaId: projectUuid.nullable(),
        contentRevisionId: projectUuid.nullable(),
        startMs: integer,
        inMs: integer,
        outMs: integer.min(1),
        durationMs: integer.min(1),
        gainDb: z.number().min(-96).max(12),
        linkedClipIds: ids,
        keyframes: z
          .array(
            z.strictObject({
              timeMs: integer,
              property: z.enum(["x", "y", "scale", "opacity", "volume"]),
              value: z.number().min(-10000).max(10000),
              interpolation: z.enum(["linear", "hold"]),
            }),
          )
          .max(1000),
      }),
    )
    .max(5000),
  transitions: z
    .array(
      z.strictObject({
        id: projectUuid,
        fromClipId: projectUuid,
        toClipId: projectUuid,
        type: z.enum(["cut", "dissolve", "fade"]),
        durationMs: integer.max(10000),
      }),
    )
    .max(1000),
});
export const rightsSchema = z.strictObject({
  id: projectUuid,
  mediaId: projectUuid,
  source: text.min(1).max(2000),
  use: text.min(1).max(2000),
  evidenceMediaIds: ids,
  state: z.enum(["unverified", "verified", "recheck"]),
  explanation: text,
});
export const issueSchema = z.strictObject({
  id: projectUuid,
  ruleId: text,
  ruleVersion: text,
  artifactId: projectUuid,
  revisionId: projectUuid,
  baselineRevisionId: projectUuid.nullable(),
  severity: z.enum(["blocking", "unknown_required", "deviation", "advice"]),
  status: z.enum([
    "open",
    "fixing",
    "recheck",
    "resolved",
    "accepted_deviation",
  ]),
  message: text,
  evidence: text,
  time: z.strictObject({ startMs: integer, endMs: integer.min(1) }).nullable(),
  method: z.enum(["local", "ai", "human"]),
  limitations: text,
});
export const renderPlanSchema = z.strictObject({
  id: projectUuid,
  compilerVersion: text,
  timelineRevisionId: projectUuid,
  timeline: timelineSchema,
  mediaHashes: z
    .array(z.strictObject({ mediaId: projectUuid, sha256: hash }))
    .max(1000),
  inputHash: hash,
  videoCodec: z.literal("h264"),
  audioCodec: z.literal("aac").nullable(),
  purpose: z.enum(["preview", "animatic", "export"]),
});
const command = <T extends z.ZodType>(payload: T) =>
  z.strictObject({
    clientOperationId: projectUuid,
    expectedRevision: integer,
    payload,
  });
const input = <T extends z.ZodType>(payload: T) => project.extend({ payload });
const write = <T extends z.ZodType>(payload: T) =>
  project.extend({ command: command(payload) });
const base = (value: Record<string, unknown>) =>
  `/api/v1/projects/${value.projectId}`;
const route = <I extends z.ZodType, O extends z.ZodType>(options: {
  method: "GET" | "POST" | "PUT";
  input: I;
  output: O;
  mutating: boolean;
  global?: boolean;
  path: (value: Record<string, unknown>) => string;
  body?: (value: z.infer<I>) => unknown;
}) => options;
const timelineId = z.strictObject({ timelineRevisionId: projectUuid });
const diagnosticOptions = z.strictObject({
  projectId: projectUuid.nullable(),
  includeProject: z.boolean(),
  includeContent: z.boolean(),
});
export const productionRoutes = {
  index: route({
    method: "GET",
    input: project.extend({ offset: integer.max(999999999).optional() }),
    output: z.strictObject({
      nextOffset: integer.nullable(),
      drafts: z
        .array(
          z.strictObject({
            id: projectUuid,
            artifactId: projectUuid,
            kind,
            savedAt: text,
          }),
        )
        .max(5000),
      adopted: z
        .array(
          z.strictObject({
            artifactId: projectUuid,
            revisionId: projectUuid,
            kind,
            confirmed: z.boolean(),
            needsUpdate: z.boolean(),
            content: object,
          }),
        )
        .max(5000),
    }),
    mutating: false,
    path: (i) => `${base(i)}/production?offset=${i.offset ?? 0}`,
  }),
  timing: route({
    method: "POST",
    input: input(
      z.strictObject({
        shotRevisionId: projectUuid,
        speechRevisionIds: ids,
        beforeMs: integer,
        afterMs: integer,
      }),
    ),
    output: z.strictObject({
      requiredMs: integer,
      plannedMs: integer,
      shortageMs: integer,
      method: z.enum(["measured", "estimated"]),
      suitable: z.boolean(),
    }),
    mutating: false,
    path: (i) => `${base(i)}/timing-checks`,
    body: (i) => i.payload,
  }),
  readiness: route({
    method: "POST",
    input: input(
      z.strictObject({
        shotRevisionId: projectUuid,
        speechRevisionIds: ids,
        capabilityId: projectUuid,
        path: z.enum(["research", "production"]),
      }),
    ),
    output: z.strictObject({
      ready: z.boolean(),
      blockers: z.array(text).max(1000),
      warnings: z.array(text).max(1000),
    }),
    mutating: false,
    path: (i) => `${base(i)}/video-readiness`,
    body: (i) => i.payload,
  }),
  rights: route({
    method: "GET",
    input: project,
    output: z.strictObject({ items: z.array(rightsSchema).max(5000) }),
    mutating: false,
    path: (i) => `${base(i)}/rights`,
  }),
  saveRights: route({
    method: "PUT",
    input: write(z.strictObject({ evidence: rightsSchema })).extend({
      evidenceId: projectUuid,
    }),
    output: receiptSchema,
    mutating: true,
    path: (i) => `${base(i)}/rights/${i.evidenceId}`,
    body: (i) => i.command,
  }),
  ducking: route({
    method: "POST",
    input: input(
      z.strictObject({
        timelineRevisionId: projectUuid,
        musicClipIds: ids,
        reductionDb: z.number().min(0).max(96),
        attackMs: integer.max(10000),
        releaseMs: integer.max(10000),
      }),
    ),
    output: z.strictObject({
      timeline: timelineSchema,
      warnings: z.array(text).max(1000),
    }),
    mutating: false,
    path: (i) => `${base(i)}/mix/ducking`,
    body: (i) => i.payload,
  }),
  timelineEdit: route({
    method: "POST",
    input: input(
      z.strictObject({
        timeline: timelineSchema,
        action: z.enum(["move", "split", "trim"]),
        clipId: projectUuid,
        startMs: integer.optional(),
        splitMs: integer.optional(),
        inMs: integer.optional(),
        outMs: integer.optional(),
        durationMs: integer.optional(),
      }),
    ),
    output: z.strictObject({
      timeline: timelineSchema,
      affectedClipIds: ids,
      subtitleSplitPreview: z
        .array(z.strictObject({ clipId: projectUuid, splitMs: integer }))
        .max(1000),
    }),
    mutating: false,
    path: (i) => `${base(i)}/timeline/edit-preview`,
    body: (i) => i.payload,
  }),
  renderPlan: route({
    method: "POST",
    input: input(timelineId.extend({ burnSubtitles: z.boolean().optional() })),
    output: renderPlanSchema,
    mutating: false,
    path: (i) => `${base(i)}/render-plan-preview`,
    body: (i) => i.payload,
  }),
  animatic: route({
    method: "POST",
    input: write(timelineId),
    output: receiptSchema,
    mutating: true,
    path: (i) => `${base(i)}/animatics`,
    body: (i) => i.command,
  }),
  replacement: route({
    method: "POST",
    input: input(
      timelineId.extend({ clipId: projectUuid, newMediaId: projectUuid }),
    ),
    output: z.strictObject({
      shortageMs: integer,
      affectedClipIds: ids,
      requiredChecks: z.array(text).max(1000),
      canPreserveEdit: z.boolean(),
    }),
    mutating: false,
    path: (i) => `${base(i)}/timeline/replacement-preview`,
    body: (i) => i.payload,
  }),
  exportFilm: route({
    method: "POST",
    input: write(
      timelineId.extend({
        targetGrantId: projectUuid,
        burnSubtitles: z.boolean(),
        overwriteConfirmed: z.boolean(),
      }),
    ),
    output: receiptSchema,
    mutating: true,
    path: (i) => `${base(i)}/exports`,
    body: (i) => i.command,
  }),
  getExport: route({
    method: "GET",
    input: project.extend({ exportId: projectUuid }),
    output: z.strictObject({
      id: projectUuid,
      timelineRevisionId: projectUuid,
      jobId: projectUuid,
      state: z.enum(["pending", "complete", "failed", "cancelled"]),
      mediaId: projectUuid.nullable(),
      inputHash: hash,
    }),
    mutating: false,
    path: (i) => `${base(i)}/exports/${i.exportId}`,
  }),
  issues: route({
    method: "GET",
    input: project.extend({
      cursor: projectUuid.nullable().optional(),
      limit: integer.min(1).max(200).optional(),
    }),
    output: z.strictObject({
      items: z.array(issueSchema).max(200),
      nextCursor: projectUuid.nullable(),
    }),
    mutating: false,
    path: (i) =>
      `${base(i)}/issues?limit=${i.limit ?? 50}${i.cursor ? `&cursor=${i.cursor}` : ""}`,
  }),
  decideIssue: route({
    method: "POST",
    input: write(
      z.strictObject({
        action: z.enum([
          "accept_deviation",
          "request_recheck",
          "attach_evidence",
        ]),
        reason: text.min(1).max(2000),
        evidenceMediaIds: ids,
      }),
    ).extend({ issueId: projectUuid }),
    output: receiptSchema,
    mutating: true,
    path: (i) => `${base(i)}/issues/${i.issueId}/decisions`,
    body: (i) => i.command,
  }),
  diagnosticPreview: route({
    method: "POST",
    input: diagnosticOptions,
    output: diagnosticOptions.extend({
      files: z.array(z.strictObject({ name: text, description: text })).max(20),
    }),
    mutating: false,
    global: true,
    path: () => "/api/v1/diagnostics/preview",
    body: (i) => i,
  }),
  diagnostic: route({
    method: "POST",
    input: z.strictObject({
      command: command(
        diagnosticOptions.extend({ targetGrantId: projectUuid }),
      ),
    }),
    output: receiptSchema,
    mutating: true,
    global: true,
    path: () => "/api/v1/diagnostics",
    body: (i) => i.command,
  }),
};
export const chooseTargetSchema = z.strictObject({
  purpose: z.enum(["export", "diagnostic"]),
});
export const chosenTargetSchema = z
  .strictObject({ grantId: projectUuid, name: text, exists: z.boolean() })
  .nullable();
export type ProductionIndex = z.infer<typeof productionRoutes.index.output>;
export type Timeline = z.infer<typeof timelineSchema>;
export type RightEvidence = z.infer<typeof rightsSchema>;
export type ProductionBridge = {
  [K in keyof typeof productionRoutes]: (
    input: z.infer<(typeof productionRoutes)[K]["input"]>,
  ) => Promise<ProjectResult<z.infer<(typeof productionRoutes)[K]["output"]>>>;
} & {
  chooseTarget(
    input: z.infer<typeof chooseTargetSchema>,
  ): Promise<ProjectResult<z.infer<typeof chosenTargetSchema>>>;
};
